Canary Checkbox V1, BFDR 2.0 pre-release Canary Cleanup

RobertScalfani opened the pull request:
This PR implements Canary Checkboxes for BFDR checkbox fields and features a cleanup of errors for the BFDR 2.0 pre-release.

- Checkboxes: Checkbox fields, such as Pregnancy Risk Factors will render as checkboxes rather than true/false. They've also had their priorities reordered so that they display in the expected order, where "grouped" sections (which cannot be programmatically grouped at this time) are together.
> Example: In Canary, navigate to "BFDR Record Inspector" and submit a record. Go to the "Pregnancy Risk Factors" section which contains a set of checkbox fields, which are now rendered as checkboxes in the expected order based on the form.
> "Checkbox" properties are determined based on if their FHIRPath includes a "section" attribute.

RobertScalfani commented:
The team walked through and discussed this version of the branch on 3/11/24.

- The library/Canary should not be enforcing the logic of only one checkbox being ticked per section. This would be handled via business rules upon record submission.
- We don't need to group the subsections (such as diabetes with prepregnancy/gestational) but they should be in order so that related fields are next to each other. This will be done using the priority field of the [Property] header.
- We will use checkboxes instead of true/false. This makes it more in-line with the worksheet and makes it possible for those fields that are unchecked to not be present in the exported JSON. Explicit falses should not have their properties exported. This needs to be checked and verified by making sure that an "inspected" record with the field present renders in the xml/json dropdowns. Will follow up with the results of this.

smacadam commented:
Notes on the Producing FHIR BFDR Records feature:
- Infections present during pregnancy split the "No Infections Present During Pregnancy" property into its on group but it should be included with the others
- The properties are hard to read because there's lots of text, could the test be shortened to just the property name and remove the code, category code, and section from the ui display?
- Final Route and Method of Delivery fields are missing, might just be resolved with a rebase
- Use the priority numbers to group risk factors together that are grouped on the form like Infertility Treatment, Artificial Insemination, and Assisted Fertilization

smacadam commented:
Test FHIR consuming
- update to use the checkboxes

RobertScalfani commented:
> Notes on the Producing FHIR BFDR Records feature:
> 
> * Infections present during pregnancy split the "No Infections Present During Pregnancy" property into its on group but it should be included with the others
> * The properties are hard to read because there's lots of text, could the test be shortened to just the property name and remove the code, category code, and section from the ui display?
> * Final Route and Method of Delivery fields are missing, might just be resolved with a rebase
> * Use the priority numbers to group risk factors together that are grouped on the form like Infertility Treatment, Artificial Insemination, and Assisted Fertilization

Addressed these. The issue with the "Final Route and Method of Delivery" fields was that they weren't present in BasicBirthRecord.json. I've added some as examples, which makes it display.

RobertScalfani commented:
> Test FHIR consuming
> 
> * update to use the checkboxes

Added this.

RobertScalfani commented:
> Previously discussed Issues for possible future reference:
> - These checkbox fields, right now, are all represented as separate fields with no connection with each other and no order. However, based on (https://www.cdc.gov/nchs/data/dvs/facility-worksheet-2016-508.pdf), There are subsections of properties that are interrelated and can have only one or no checks ticked. For instance, in Pregnancy Risk Factors, there is a Diabetes subsection in which Pregnancy Diabetes OR Gestational diabetes can be checked. As it's represented now, these two properties are not even one-after-another, let alone linked in a Diabetes category. I'm not sure that there is any way to link to these together because there doesn't seem to be a Diabetes category represented in the code for these properties. Is this a limitation of the IG? There are some "code" and "categoryCode" properties in the FHIRPath, but there are not unique between the sub-subcategories.
> - In an ideal scenario, the properties within these subsections could be radio-groups instead of checkboxes to enforce only one being able to be clicked.

RobertScalfani commented:
This is now ready for review, it includes checkboxes and the BFDR 2.0 pre-release cleanup.
